### PR TITLE
Add methods for piecewise exponential distributions

### PR DESCRIPTION
**Pull request by mclements**
Devin,

As a start, here are some modest changes for piecewise exponential distributions.

Please shout out if anything is unclear.

Kindly, Mark.

### CODE REVIEW

File: inst/include/hesim/utils.h [hunk 1]
```diff
@@ -293,6 +296,28 @@ class trans_mat {
   
 };
 
+  /***************************************************************************/
+  /**
+   * Integer bound for the x being in the vector range such that x<=range[1] -> 0
+   * and range[i] < x <= range[i+1] -> i
+   */
```
Comment (dincerti) [inline inst/include/hesim/utils.h:303]: It's not immediately obvious what the function does so I would document the parameters and the returns as well.

File: tests/testthat/test-cpp-distributions.R [hunk 2]
```diff
@@ -51,19 +51,21 @@ test_that("pwexp", {
   pwexp <- new(PwExp, rate = rate, time = time)
   
   # pdf
-  expect_error(pwexp$pdf(3))
+  expect_equal(pwexp$pdf(5.5), 1.2*exp(-(5.0*0.8 + 0.5*1.2)))
```
Comment (dincerti) [inline tests/testthat/test-cpp-distributions.R:54]: We should test the edge case and confirm that we get the same answer as with `msm`:
```suggestion
  expect_equal(pwexp$pdf(5.5), 1.2*exp(-(5.0*0.8 + 0.5*1.2)))
  
  expect_equal(pwexp$pdf(5), 1.2*exp(-(5.0*0.8)))
  expect_equal(pwexp$pdf(5), 1.2*exp(-(5.0*0.8)))
```

File: tests/testthat/test-cpp-distributions.R [hunk 3]
```diff
@@ -51,19 +51,21 @@ test_that("pwexp", {
...
   
   # cdf
-  expect_error(pwexp$cdf(3)) 
+  expect_equal(pwexp$cdf(5.5), 1.0-exp(-(5.0*0.8 + 0.5*1.2)))
   
   # quantile
-  expect_error(exp$quantile(.025))
+  expect_equal(pwexp$quantile(1.0-exp(-(5.0*0.8 + 0.5*1.2))), 5.5)
```
Comment (dincerti) [inline tests/testthat/test-cpp-distributions.R:60]: You should test the CDF and quantile function against `msm` too.

**Review by dincerti — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments on function names and the edge case.

File: inst/include/hesim/utils.h [hunk 4]
```diff
@@ -293,6 +296,28 @@ class trans_mat {
...
+  inline int hesim_bound(double x, std::vector<double> range) {
+    auto lower = std::lower_bound(range.begin(), range.end(), x);
+    if (lower==range.end() || (*lower >= x && x > range[0]))
+      lower--;
+    return std::distance(range.begin(), lower);
+  }
```
[the remark this answers is not in the document]
Comment (mclements) [reply inst/include/hesim/utils.h:309]: I have refactored the implementation to use `findInterval` from the C-API and renamed the function to `find_interval`.

File: inst/include/hesim/utils.h [hunk 5]
```diff
@@ -293,6 +296,28 @@ class trans_mat {
...
+
+  /***************************************************************************/
+  /**
+   * Test whether the value x is in the vector/set lookup
+   */
+  template<class T, class V>
+  bool member_of(T x, V lookup) {
```
[the remark this answers is not in the document]
Comment (mclements) [reply inst/include/hesim/utils.h:316]: Now changed.

File: tests/testthat/test-cpp-utils.R [hunk 6]
```diff
@@ -37,4 +37,27 @@ test_that("max_lt", {
...
+
+# Test c++ function hesim_bound -----------------------------------------------------
+test_that("hesim_bound", {
+    range = c(0,3,5)
+    expect_equal(hesim:::C_test_hesim_bound(-1,range),0)
+    expect_equal(hesim:::C_test_hesim_bound(0,range),0)
+    expect_equal(hesim:::C_test_hesim_bound(1,range),0)
+    expect_equal(hesim:::C_test_hesim_bound(3,range),0) # should this be 1?
+    expect_equal(hesim:::C_test_hesim_bound(3.1,range),1)
+    expect_equal(hesim:::C_test_hesim_bound(5,range),1) # should this be 2?
```
[the remark this answers is not in the document]
Comment (mclements) [reply tests/testthat/test-cpp-utils.R:50]: Good - I also prefer that the left side of an interval is closed. Now changed.

File: inst/include/hesim/utils.h (hunk 1 above)
Comment (mclements) [reply inst/include/hesim/utils.h:303]: Now documented.

File: tests/testthat/test-cpp-distributions.R (hunk 3 above)
Comment (mclements) [reply tests/testthat/test-cpp-distributions.R:63]: Now tested against `msm`.

**Comment by mclements** [comment]
Devin: I have also added some casts in a number of integer comparisons to address some long-standing warnings.

**Review by dincerti — APPROVED** [review]
